engine/table/src/main/java/io/deephaven/engine/table/impl/TableWithDefaults.java

```diff
@@ -1113,6 +1113,110 @@ default Table rollup(Collection<? extends Aggregation> aggregations, boolean inc
         return rollup(aggregations, includeConstituents, ZERO_LENGTH_SELECTABLE_ARRAY);
     }
 
+
+    // -----------------------------------------------------------------------------------------------------------------
+    // UpdateBy Operations
+    // -----------------------------------------------------------------------------------------------------------------
+
+    /**
+     * <p>
+     * Create a table with the same {@link #getRowSet() rowSet} as it's parent that will perform the specified set of
+     * row based operations to it. As opposed to {@link #update(String...)} these operations are more restricted but are
+     * capable of processing state between rows. This operation will group the table by the specified set of keys if
+     * provided before applying the operation.
+     * </p>
+     *
+     * @param operations the operations to apply to the table.
+     * @param byColumns the columns to group by before applying.
+     * @return a table with the same index, with the specified operations applied to each group defined by the
+     *         {@code byColumns}
+     */
+    @ConcurrentMethod
+    Table updateBy(@NotNull final UpdateByControl control, @NotNull final Collection<UpdateByClause> operations,
+            final MatchPair... byColumns);
```

Comment:
If these are true MatchPair, we might prefer to use the type `Collection<? extends Selectable> groupByColumns`?